Work in progress

ziir opened the pull request:
```
gandi paas create --type nodejspgsql --name test-paas-ux
```

Either clone your newly created instance's default vhost :

```
gandi paas clone test-paas-ux [--origin gandi]
```

Or, create a new local repository for your project :

```
mkdir test-paas-ux
cd test-paas-ux
git init
```

Unless you are starting from an existing repository ?

```
cd test-paas-ux
```

Anyway, now is the time to add the `gandi` git remote to this local repository, unless you cloned it of course.

```
gandi paas attach test-paas-ux [--remote gandi]
```

Which will print the following :

```
Added remote `gandi` to your local git repository.
Use `git push gandi master` to push your code to the instance.
Then `$ gandi deploy` to build and deploy your application.
```

Write / fetch some nodejs app code into your directory :)
Then, simply `git add`, `git commit` and `git push` to the `gandi` remote :

```
git push gandi master
```

Now that your code has been pushed to your instance's default vhost git repository, time for deploy 

(moving the code from the git repository to the instance's vhost itself - `/srv/data/web/vhosts/default/`)

```
gandi deploy
```

And you're done! :)

kalou committed:
simplify gandi vhost create

Part of discussed ux updates

It seems more natural to use :

$ gandi vhost create paas0.tfz.net --paas ctnr3 --alter-zone

than to specify the vhost with --vhost

Conflicts:
	gandi/cli/commands/vhost.py

Commit:
Introduce gandi paas attach command

Commit:
Introduce Paas.git_remote method

Commit:
Refactor gandi paas clone command

gandi paas clone command now requires the paas instance name and accepts a destination --directory option.

Commit:
Simplify gandi deploy

MoiTux commented on gandi/cli/commands/paas.py:
```diff
@@ -85,39 +85,44 @@ def info(gandi, resource, stat):
-        gandi.error('missing VHOST parameter')
+def attach(gandi, name, vhost, remote):
+    """Add remote for an instance's default vhost to the local git """
+    """repository."""
```
```
"""Add remote ....
repository.
"""
```

ziir replied:
Ok

MoiTux commented on gandi/cli/modules/paas.py:
```diff
@@ -29,6 +32,70 @@ def type_list(cls, options=None):
+            cls.echo('Adding remote `%s` to your local git repository.'
+                     % (remote_name))
+            cls.echo('Use `git push %s master` to push your code to the '
+                     'instance.' % (remote_name))
```
name ?

MoiTux replied:
s/remote_name/name

ziir replied:
I don't think so.
I believe we wish to have :

```
cd mon-paas
gandi paas attach mon-paas
git push gandi master
```

Or

```
cd mon-paas
gandi paas attach mon-paas --remote origin
git push origin master
```

MoiTux replied:
yes indeed I read it like that 'Use `git push %s master` to push your code to the instance: %s' thus using name instead of remote_name

MoiTux commented on gandi/cli/tests/commands/test_paas.py:
```diff
@@ -190,48 +190,30 @@ def test_clone_no_conf(self):
-        with mock.patch('gandi.cli.modules.vhost.os.chdir',
-                        create=True) as mock_chdir:
-            mock_chdir.return_value = mock.MagicMock()
+#     def test_deploy_no_conf_vhost(self):
```
to remove avoid using comment

ziir replied:
Yes, I don't think we want to allow deploy without a configuration file anyway.

sayoun commented on gandi/cli/commands/vhost.py:
```diff
@@ -87,9 +86,10 @@ def info(gandi, resource, id):
-def create(gandi, vhost, paas, ssl, private_key, alter_zone, poll_cert,
-           background):
+def create(gandi, paas, ssl, private_key, alter_zone, poll_cert,
+           background, vhost):
```
you don't need to change the method signature as the parameters are passed using kwargs by click

sayoun commented on gandi/cli/modules/paas.py:
```diff
@@ -29,6 +32,67 @@ def type_list(cls, options=None):
+        """Attach an instance's vhost to a remote from the local repository."""
+        remote_url = cls.git_remote(name, vhost)
+
+        ret = cls.execute('git remote add %s %s' % (remote_name, remote_url,))
```
are you sure you are in a valid git repository when running this command ? Maybe you should handle this type of error to display a proper message

ziir replied:
I'm not, but I figured I'd let this responsibility to git itself, at least for the moment.
If I were to change my mind about this, it'd mean handle :
- current directory is not a git repository :

```
fatal: remote gandi already exists.
```
- remote already exists

```
fatal: Not a git repository
```

Which is fine by me, I'm just not sure how to check which error case I'm in.
What do you think ?

sayoun replied:
I thought that it could raise an ugly TB in case the command failed but instead it displayed the git error message, which is fine indeed. So you can leave it as it is.

Commit:
Allow only 1vhost in paas commands; Refactoring

Also, remove Paas.init_conf method
Use Paas.save_config() and Vhost.create() independently instead

Commit:
Remove Vhost.init_vhost method

Use Paas.save_config instead

Commit:
Add support --vhost support for gandi paas clone

Commit:
Rely on GIT error message only for clone errors

Commit:
Move deploy code from command to module

Also, remove the --vhost option.
Always deploy by trusting current directory
local configuration file that was created either from:

gandi paas clone mypaas --vhost my.vhost

or

gandi paas attach mypaas --vhost my.vhost

Commit:
Fixup: remove git_remote; paas attach; paas clone

Commit:
Fixup: on paas create, call paas.info if --vhost

One less paas.info useless call if no --vhost option given.
Also, fixes paas test test_create_default

Commit:
Introduce Paas.save_config() method
